### DISCUSSION TITLE
Django Logging with Gunicorn

### DISCUSSION THREAD

**Discussion by phucth102** [miguelgrinberg/python-socketio#770] [Q&A]
In my project,

I have built the django webapp and tracking log via gunicorn. After that I add socketio, it's run well, I can track log from socketio. However, I can not capture logs of django app such as request url, print(), ...
I guess Gunicorn recognized Socket server as the main server, it skip Django app.

Could you help me find some ways to get Django logs.
Thanks a lot.

```python
# wsgi.py
import os
from django.core.wsgi import get_wsgi_application
import socketio
async_mode = 'gevent'
sio = socketio.Server(async_mode=async_mode, cors_allowed_origins='*', logger=False, engineio_logger=False)
os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'core.settings')
application = get_wsgi_application()
application = socketio.WSGIApp(sio, application)
```
```python
# gunicorn-cfg.py
bind = '0.0.0.0:5005'
workers = 1
worker_connections=1000
loglevel = 'debug'
capture_output = True
enable_stdio_inheritance = True
worker_class = 'geventwebsocket.gunicorn.workers.GeventWebSocketWorker'
```

```# log example before integrating socketIO
core-app_1  | [2021-08-19 19:34:10 +0000] [10] [DEBUG] GET /login/
core-app_1  | [2021-08-19 19:34:12 +0000] [10] [DEBUG] GET /admin/
core-app_1  | [2021-08-19 19:34:12 +0000] [10] [DEBUG] GET /static/admin/css/base.css
core-app_1  | [2021-08-19 19:34:12 +0000] [10] [DEBUG] GET /static/admin/css/datetime_widget.css
core-app_1  | [2021-08-19 19:34:12 +0000] [10] [DEBUG] GET /static/admin/css/dashboard.css
core-app_1  | [2021-08-19 19:34:12 +0000] [10] [DEBUG] GET /static/admin/js/vendor/jquery/jquery.js
```

**Comment by miguelgrinberg** [comment]
I think this is because you are not using the Django server anymore, right? You switched to gevent, which is a different web server. Have you seen the `log_output` option that you pass to the Socket.IO object? That may help.

**Reply by phucth102** [reply]
Thanks @miguelgrinberg,
I still use Django app because I have sent request to Django urls. I deployed following document at https://python-socketio.readthedocs.io/en/latest/server.html#gevent-with-gunicorn
I have no see any option `log_output` from SocketIO object. You mean is ```socketio.Server()```, right?

**Reply by miguelgrinberg** [reply]
Yes, you are using your Django app, but you are running it on a different web server. The logs that you are used to see are printed by the Django dev server. You now switched to the gunicorn/gevent web server, which has different logging options. I mentioned the `log_output` option because I assumed you were using gevent standalone. Since you are using it along with gunicorn, the logging is controlled by Gunicorn, so see the command line option in Gunicorn to log requests.

**Reply by phucth102** [reply]
Thanks @miguelgrinberg for your support.
I decided to use LOGGING option in setting.py of Django app and I was managed capture logs.
However, I have a new problem, socketIO client automatically reloads continuously.

```python
LOGGING = {
    'version': 1,
    'disable_existing_loggers': False, 
    'formatters': {
        'verbose': {
            'format': '{name} {levelname} {asctime} {module} {process:d} {thread:d} {message}',
            'style': '{',
        },
        'simple': {
            'format': '{levelname} {message}',
            'style': '{',
        },
    },
    'handlers': {
        'file': {
            'class': 'logging.FileHandler',
            'filename': 'logs/backend.log',
            'formatter': 'simple',
        }
    },
    'loggers': {
        '': {
            'level': 'INFO',
            'handlers': ['file'],
        },
    },
}
```

``` javascript
<!DOCTYPE html>
<head>
    <script src="https://cdn.socket.io/3.1.3/socket.io.min.js" integrity="sha384-cPwlPLvBTa3sKAgddT6krw0cJat7egBga3DJepJyrLl4Q9/5WLra3rrnMcyTyOnh" crossorigin="*"></script>
</head>
<body>
    <script>
        const socket = io("http://localhost:5005");
        console.log(socket)
        socket.on('connect', message => {
            console.log(message)
        })
        socket.on('message', message => {
            console.log(message)
        })
    </script>
</body>
```

I can not find out any solution for this circumstance.

**Reply by miguelgrinberg** [reply]
I don't know what "socketIO client automatically reloads continuously" means in detail. Also note that the socket.io client that you are using is not a project of mine.

**Reply by phucth102** [reply]
Hi @miguelgrinberg,
I resolved that issue. The socket.io client was changed to socket `socket.io-client` (from npm) and it worked so good.
Thanks a lot.
